components/authz-service/storage/postgres/migration/sql/61_use_role_db_id_in_iam_statements.up.sql

```diff
@@ -0,0 +1,233 @@
+                        ),
+                        (
+                            SELECT
+                                COALESCE(json_agg(proj.id) FILTER (WHERE proj.id IS NOT NULL),
+                                    '[]')
+                                FROM iam_statement_projects AS stmt_projs
+                            LEFT OUTER JOIN iam_projects AS proj ON stmt_projs.statement_id = stmt.db_id WHERE stmt_projs.project_id = proj.db_id) AS projects FROM iam_statements stmt
+                    INNER JOIN iam_policies ON stmt.policy_id = pol.db_id
+                GROUP BY
+                    stmt.db_id,
+                    stmt.id,
+                    stmt.effect,
+                    stmt.actions,
+                    stmt.resources,
+                    stmt.role_id
+)
+            SELECT
+                array_agg(statement_rows) FILTER (WHERE statement_rows.id IS NOT NULL)
+                FROM statement_rows) AS statements,
+            (
+            SELECT
+                array_agg(mem) FILTER (WHERE mem.id IS NOT NULL)
+                FROM iam_policy_members AS pol_mems
+            LEFT OUTER JOIN iam_members AS mem ON pol_mems.member_id = mem.db_id WHERE pol_mems.policy_id = pol.db_id) AS members,
+    (
+    SELECT
+        array_agg(proj.id) FILTER (WHERE proj.id IS NOT NULL)
+        FROM iam_policy_projects AS pol_projs
+        LEFT OUTER JOIN iam_projects AS proj ON pol_projs.project_id = proj.db_id WHERE pol_projs.policy_id = pol.db_id) AS projects FROM iam_policies AS pol
+GROUP BY
+    pol.db_id,
+    pol.id,
+    pol.name,
+    pol.type
+)
+SELECT
+    json_build_object('id', temp.id, 'name', temp.name, 'type', temp.type, 'statements', COALESCE(temp.statements, '{}'), 'members', COALESCE(temp.members, '{}'), 'projects', COALESCE(temp.projects, '{}')) AS POLICY
+FROM
+    temp
+WHERE
+    projects_match (temp.projects::TEXT[], _projects_filter);
+$$
+LANGUAGE sql;
+
+CREATE OR REPLACE FUNCTION
+    insert_iam_statement_into_policy(_policy_id TEXT, _statement_id UUID, _statement_effect iam_effect, _statement_actions TEXT[],
+    _statement_resources TEXT[], _statement_role TEXT, _statement_projects TEXT[])
+        RETURNS void AS $$
+            BEGIN
+                -- if NULL or an empty string was passed for the role, we shouldn't try to insert a role.
+                IF _statement_role IS NULL OR _statement_role=''
+                THEN
+                    INSERT INTO iam_statements (policy_id, id, effect, actions, resources)
+                        VALUES (policy_db_id(_policy_id), _statement_id, _statement_effect, _statement_actions, _statement_resources);
+
+                -- otherwise, we should try to insert a role. however, we want to catch the case where role_db_id returns NULL.
+                -- if we don't then the insert will just insert NULL for role_id, which is not what we want if a role was passed in.
+                ELSE
+                    IF role_db_id(_statement_role) IS NULL
```

Comment:
would using `STRICT` work here?

Reply:
If the function `role_db_id` was using `SELECT db_id STRICT INTO ...`, this case wouldn't be possible -- so, we might _want it to be possible_? 🤔

Reply:
I'm thinking leave it for now since all the other ones don't use strict either and it's used in the tests a bunch.

Reply:
Ah looks like it's inconsistent :S I'm gonna add a subtask to make these consistent.

Strict:

```
statement_db_id
```

Not Strict:

```
role_db_id
project_db_id
policy_db_id
```

Reply:
Added subtask to https://github.com/chef/automate/issues/832

Reply:
> Ah looks like it's inconsistent :S

Yup! Just learnt how to do that halfway into the work. ;)

Reply:
also not strict: `member_db_id`

Reply:
strict (as opposed to what's claimed above): `policy_db_id` (as redefined in migration 55)